src/rt/sections_osx_x86_64.d

```diff
@@ -114,11 +113,62 @@ void scanTLSRanges(void[] rng, scope void delegate(void* pbeg, void* pend) nothr
 
 private:
 
-extern(C) void _d_dyld_getTLSRange(void*, void**, size_t*);
+// Declarations from dyld_priv.h in dyld, available on 10.7+.
+enum dyld_tlv_states
+{
+    allocated = 10,
+    deallocated = 20
+}
+
+struct dyld_tlv_info
+{
+    size_t info_size;
+    void * tlv_addr;
+    size_t tlv_size;
+}
+
+alias dyld_tlv_state_change_handler = Block!(void, dyld_tlv_states, dyld_tlv_info*)*;
+extern(C) void dyld_enumerate_tlv_storage(dyld_tlv_state_change_handler handler);
 
-__gshared SectionGroup _sections;
 ubyte dummyTlsSymbol;
 
+struct TLSRange
+{
+    void* start;
+    size_t size;
+
+    bool isValid()
+    {
+        return start !is null && size > 0;
+    }
+
+    void[] toArray()
+    {
+        return start[0 .. size];
+    }
+}
+
+TLSRange getTLSRange()
+{
+    void* tlsSymbol = &dummyTlsSymbol;
+    TLSRange range;
+
+    auto handler = block((dyld_tlv_states state, dyld_tlv_info* info){
+        assert(state == dyld_tlv_states.allocated);
+        if (info.tlv_addr <= tlsSymbol && tlsSymbol <
+            (info.tlv_addr + info.tlv_size)
+        )
+        {
+            range = TLSRange(info.tlv_addr, info.tlv_size);
+        }
+    });
+
+    dyld_enumerate_tlv_storage(&handler);
```

Comment:
> Is that possible to do on the call site?

Yes. Either you can use:
```D
scope dg = (dyled_tlv_states state, dyld_tlv_info* info) { /*...  */ }
auto handler = block(dg);
```

You could change the definition of `block` to take a parameter of type `scope R delegate(Params) dg` which will mean delegate literals passed directly will never allocate. But see the next point.

> What do you mean with "true delegate"? A closure that captures local variables which can be used after a function returns? Yes.

Quite a poor phrasing on my side, the question was actually "can the block escape the lexical scope in which it was declared", to which the answers seems to be yes (and so you can't use `scope delegate` for `block`'s prototype).